### PR TITLE
Fixed empty block handling in concurrent_processing branch

### PR DESCRIPTION
**Pull request by Danstahr**
Makes all the tests pass!

### CODE REVIEW

**Comment by kenkendk** [comment]
Wow!

File: Duplicati/Library/Main/Database/LocalDatabase.cs [hunk 1]
```diff
@@ -955,40 +960,41 @@ System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
...
-            AND ""G"".""Index"" = 0
-            AND ""I"".""BlocksetID"" = ""E"".""ID""
-            AND ""I"".""BlockID"" = ""H"".""ID""
-            AND ""I"".""Index"" = 0
+          ""A"".""BlocksetId"" >= 0 AND
+          ""D"".""FilesetID"" = ? AND
+          (""I"".""Index"" = 0 OR ""I"".""Index"" IS NULL) AND  
+          (""G"".""Index"" = 0 OR ""G"".""Index"" IS NULL)
         ) J
     LEFT OUTER JOIN
```
Comment (kenkendk) [inline Duplicati/Library/Main/Database/LocalDatabase.cs]: You have rewritten the crazy join in a cartesion-product-style syntax with left-joins?

How sure are you that your rewrite is correct?
I find it hard to follow the rewrite :/

File: Duplicati/Library/Main/Operation/RestoreHandler.cs [hunk 2]
```diff
@@ -778,8 +784,6 @@ private static void PrepareBlockAndFileList(LocalRestoreDatabase database, Optio
                 {
                     // Find the largest common prefix
                     var largest_prefix = options.DontCompressRestorePaths ? "" : database.GetLargestPrefix();
-                    if (options.DontCompressRestorePaths)
```
Comment (kenkendk) [inline Duplicati/Library/Main/Operation/RestoreHandler.cs]: Dooh!

File: Duplicati/Library/Main/Operation/Backup/StreamBlockSplitter.cs [hunk 3]
```diff
@@ -103,13 +103,10 @@ public static Task Run(Options options, BackupDatabase database, ITaskReader tas
                             var lastread = 0;
                             var buf = new byte[blocksize];
                             var lastupdate = DateTime.Now;
-                            var firstBlock = true;
 
                             // Core processing loop, read blocks of data and hash individually
-                            while (((lastread = await stream.ForceStreamReadAsync(buf, blocksize)) != 0) || firstBlock)
+                            while (((lastread = await stream.ForceStreamReadAsync(buf, blocksize)) != 0))
```
Comment (kenkendk) [inline Duplicati/Library/Main/Operation/Backup/StreamBlockSplitter.cs]: The `firstblock` was used to ensure that at least one block was emitted, also for empty files.
Now that it is removed, it is possible to have files with zero blocks, and I could imagine that one or more queries fail because of this assumption.

Not sure how to find these though...

File: Duplicati/Library/Main/Operation/RestoreHandler.cs (hunk 2 above)
Comment (Danstahr) [inline Duplicati/Library/Main/Operation/RestoreHandler.cs]: This is not a big deal, just a tiny bit of redundant work harming readability.

Comment (kenkendk) [inline Duplicati/Library/Main/Operation/RestoreHandler.cs]: I know... the "Dooh!" was me seeing I wrote stupid code :)

File: Duplicati/Library/Main/Operation/Backup/StreamBlockSplitter.cs (hunk 3 above)
Comment (Danstahr) [inline Duplicati/Library/Main/Operation/Backup/StreamBlockSplitter.cs]: I only found one query that was failing because of this - the crazy join you commented on above. I did a bit of manual testing and ran all unit tests with positive results. 

I'm still not 100% convinced that nothing blows up with this change in corner cases, but we'll need to test this branch thoroughly before making it available to the general public anyway.

File: Duplicati/Library/Main/Database/LocalDatabase.cs (hunk 1 above)
Comment (Danstahr) [inline Duplicati/Library/Main/Database/LocalDatabase.cs]: Frankly, I often find Duplicati SQL queries hard to follow :).

The rewrite is equivalent to the previous version with the following exceptions:
- Previously it used inner joins, now we're using left joins. This is needed because we might not have a block for all entries anymore. I could combine comma joins with explicit left joins but I find such queries even more cryptic.
 - Added a condition for non-negative blockset IDs. This is to filter out special blockset values (currently -100 and -200).
 - Added IS NULL conditions for possible indices. The purpose of the where clause is to always pick the first block, but what if we don't have any blocks -> allow null as well.

Comment (kenkendk) [inline Duplicati/Library/Main/Database/LocalDatabase.cs]: I agree that the SQL queries are hard to follow!

Only way I can think of that makes it easier is to re-organize the data in a non-normalized form, but that would have other downsides, including increased database size :(

**Comment by Danstahr** [comment]
The tests on Travis are flaky. What's annoying is they're flaky **only** on Travis and I'm not able to reproduce the failures locally, which makes debugging pain in the butt. 

My first guess is a race condition or a stream not being properly closed *somewhere*.

**Comment by Danstahr** [comment]
I managed to get the tests consistently passing. The fix is hacky, but does the trick for now.
